Week_1/P1.cpp

```diff
@@ -0,0 +1,39 @@
+class Foo {
+
+private:
+    mutex gMutex;
+    condition_variable gCv;
+    int currTurn = 1;
+    bool isPrinting = false; 
+
+    void printsWrapper(function<void()> printFunction, int turn) {
+        {
+            unique_lock<mutex> lock(gMutex);
+            gCv.wait(lock, [this, turn]{ return (turn == currTurn) && !isPrinting; });
+            isPrinting = true;
+        }
+        printFunction();
+        {
+            unique_lock<mutex> lock(gMutex);
+            ++currTurn;
+            isPrinting = false;
+            gCv.notify_all();
+        }
+    }
+
+public:
+    Foo() {
+    }
+
+    void first(function<void()> printFirst) {
+        printsWrapper(printFirst, 1);
+    }
+
+    void second(function<void()> printSecond) {
+        printsWrapper(printSecond, 2);
+    }
+
+    void third(function<void()> printThird) {
+        printsWrapper(printThird, 3);
+    }
+};
```

Comment:
Add here a proof of correctness in a long comment.  I'll follow up with feedback once others submit the code and peer-review.